subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java

```diff
@@ -105,13 +106,15 @@ public void directoryContentObserved(File file, String consumer) {
 
     @NonNullApi
     public static class CallInterceptorRegistry {
-        private static final AtomicBoolean INTERCEPTORS_LOADED = new AtomicBoolean();
+        private static final Map<ClassLoader, Void> LOADED_FROM_CLASSLOADERS = Collections.synchronizedMap(new WeakHashMap<>());
         private static volatile CallSiteDecorator currentGroovyCallDecorator = new CallInterceptorsSet(GroovyCallInterceptorsProvider.DEFAULT);
         private static volatile JvmBytecodeInterceptorSet currentJvmBytecodeInterceptors = JvmBytecodeInterceptorSet.DEFAULT;
 
         public synchronized static void loadCallInterceptors(ClassLoader classLoader) {
-            if (INTERCEPTORS_LOADED.getAndSet(true)) {
-                return;
+            if (LOADED_FROM_CLASSLOADERS.containsKey(classLoader)) {
```

Comment:
:nail_care: can be simplified to `if (LOADED_FROM_CLASSLOADERS.put(classLoader, null) != null) { throw ... }`, but that's mostly a matter of taste, I don't insist.